### PR TITLE
Modifying metadata registries

### PR DESCRIPTION
**Pull request by Atmire-Kristof**
This PR depends on the REST PR https://github.com/DSpace/DSpace/pull/2291

The general purpose of this PR is to expand the currently existing metadata schema and field registry pages allowing admins to create, edit and delete schemas and fields.

**URLS**
To modify schemas, visit `[base-url]/admin/registries/metadata`

Here, you can create a new schema by filling in the form at the top of the list and clicking “Submit” (both fields are required).
To edit a schema, click one of the existing ones in the list at the bottom. It will automatically fill in the form with the current values.
To delete schemas, select one or multiple by selecting the checkboxes in the list and clicking “Delete selected”.
Selected schemas will be remembered when navigating further by using the Store.

To modify fields, either click one of the schemas on the page above, or visit `[base-url]/admin/registries/metadata/[schema-prefix]`

Modifying fields happens in the exact same way as you would modify schemas.
Only the element input is required to add/edit a field, the rest is optional.

**Authentication**
An AuthenticatedGuard is being used for all the /admin routes.

**Notes**
This PR is dependent on the same response-cache refactoring as for example #350 and #348, which might make it seem there’s more changes here than necessary.

### CODE REVIEW

**Comment by paulo-graca** [comment]
Does this Pull Request depends on other DSpace PR?
I'm getting an «Http failure response for http://localhost/spring-rest/api/core/metadatafields?schemaId=6: 404 Not Found» Error

**Comment by artlowel** [comment]
@paulo-graca Yes, it depends on https://github.com/DSpace/DSpace/pull/2291. I'll add that to the PR description.

**Comment by paulo-graca** [comment]
This image https://drive.google.com/file/d/16rym5Vrx1sidYhWZyJvdgLrX7jdE7v5x/view shows what I was referring when said "The fields list is visible (without the actions) even when your logged session expires"

File: src/app/+admin/admin-registries/metadata-registry/metadata-schema-form/metadata-schema-form.component.ts [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+  /**
+   * An EventEmitter that's fired whenever the form is being submitted
+   */
+  @Output() submitForm: EventEmitter<any> = new EventEmitter();
+
+  constructor(public registryService: RegistryService, private formBuilderService: FormBuilderService, private translateService: TranslateService) {
+  }
+
+  ngOnInit() {
```
[the remark this answers is not in the document]
Comment (artlowel) [reply src/app/+admin/admin-registries/metadata-registry/metadata-schema-form/metadata-schema-form.component.ts]: This should be fixed now

File: src/app/+admin/admin-registries/metadata-schema/metadata-field-form/metadata-field-form.component.ts [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+   */
+  @Output() submitForm: EventEmitter<any> = new EventEmitter();
+
+  constructor(public registryService: RegistryService,
+              private formBuilderService: FormBuilderService,
+              private translateService: TranslateService) {
+  }
+
+  ngOnInit() {
+    combineLatest(
```
[the remark this answers is not in the document]
Comment (artlowel) [reply src/app/+admin/admin-registries/metadata-schema/metadata-field-form/metadata-field-form.component.ts]: This should be fixed now

**Comment by artlowel** [comment]
@paulo-graca I agree with you on those usability issues. For this first iteration we made something that leans very close to how it worked in dspace 6 xmlui, but of course there was no editing in dspace 6, so what was there was has been modified slightly to allow for editing. 

Rather than come up with an approach for this feature separately, I suggest we wait for the Admin Item Edit PR, because the functionality is very similar. We compare what we like and what we don't in each PR, and decide which approach we want to use, and whether or not we change it now, or after the preview release.

File: src/app/+admin/admin-registries/metadata-registry/metadata-registry.reducers.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+ * The initial state.
+ */
+const initialState: MetadataRegistryState = {
+  editSchema: null,
+  selectedSchemas: [],
+  editField: null,
+  selectedFields: []
+};
+
+export function metadataRegistryReducer(state = initialState, action: MetadataRegistryAction): MetadataRegistryState {
```
Comment (tdonohue) [inline src/app/+admin/admin-registries/metadata-registry/metadata-registry.reducers.ts]: Please add TypeDocs to describe this function

File: src/app/+admin/admin-registries/metadata-schema/metadata-field-form/metadata-field-form.component.ts [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+   * An EventEmitter that's fired whenever the form is being submitted
+   */
+  @Output() submitForm: EventEmitter<any> = new EventEmitter();
+
+  constructor(public registryService: RegistryService,
+              private formBuilderService: FormBuilderService,
+              private translateService: TranslateService) {
+  }
+
+  ngOnInit() {
```
Comment (tdonohue) [inline src/app/+admin/admin-registries/metadata-schema/metadata-field-form/metadata-field-form.component.ts]: Like you've done with `onCancel` and `onSubmit`, it'd be good to add a basic description of what `ngOnInit()` is doing here, as it's rather large.

File: src/app/core/cache/response.models.ts [hunk 5]
```diff
@@ -72,6 +72,15 @@ export class MetadataschemaSuccessResponse extends RestResponse {
   }
 }
 
+export class MetadatafieldSuccessResponse extends RestResponse {
```
Comment (tdonohue) [inline src/app/core/cache/response.models.ts]: TypeDocs on this class, please

File: src/app/core/data/metadata-schema-data.service.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import { ObjectCacheService } from '../cache/object-cache.service';
+import { MetadataSchema } from '../metadata/metadataschema.model';
+import { NormalizedMetadataSchema } from '../metadata/normalized-metadata-schema.model';
+import { NotificationsService } from '../../shared/notifications/notifications.service';
+import { HttpClient } from '@angular/common/http';
+import { ChangeAnalyzer } from './change-analyzer';
+import { NormalizedObjectBuildService } from '../cache/builders/normalized-object-build.service';
+
+@Injectable()
+export class MetadataSchemaDataService extends DataService<NormalizedMetadataSchema, MetadataSchema> {
```
Comment (tdonohue) [inline src/app/core/data/metadata-schema-data.service.ts]: TypeDocs on this class, please

File: src/app/core/data/metadatafield-parsing.service.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
+import { DSpaceRESTv2Serializer } from '../dspace-rest-v2/dspace-rest-v2.serializer';
+import { DSpaceRESTV2Response } from '../dspace-rest-v2/dspace-rest-v2-response.model';
+import { RestRequest } from './request.models';
+import { ResponseParsingService } from './parsing.service';
+import { Injectable } from '@angular/core';
+import { MetadatafieldSuccessResponse, MetadataschemaSuccessResponse, RestResponse } from '../cache/response.models';
+import { MetadataField } from '../metadata/metadatafield.model';
+
+@Injectable()
+export class MetadatafieldParsingService implements ResponseParsingService {
```
Comment (tdonohue) [inline src/app/core/data/metadatafield-parsing.service.ts]: TypeDocs on this class please.  Also, the `parse` method below could use a brief description

File: src/app/core/data/request.models.ts [hunk 8]
```diff
@@ -217,6 +220,46 @@ export class IntegrationRequest extends GetRequest {
   }
 }
 
+export class CreateMetadataSchemaRequest extends PostRequest {
```
Comment (tdonohue) [inline src/app/core/data/request.models.ts]: A basic description or Typedocs should be added for all these new `*Request` classes in this file.  (See below as well)

File: src/app/core/data/request.reducer.ts [hunk 9]
```diff
@@ -105,3 +109,13 @@ function resetResponseTimestamps(state: RequestState, action: ResetResponseTimes
   });
   return newState;
 }
+
+function removeRequest(state: RequestState, action: RequestRemoveAction): RequestState {
```
Comment (tdonohue) [inline src/app/core/data/request.reducer.ts]: Typedocs please

File: src/app/core/data/request.service.ts [hunk 10]
```diff
@@ -54,6 +55,25 @@ export class RequestService {
     return pathSelector<CoreState, string>(coreSelector, 'index', IndexName.UUID_MAPPING, uuid);
   }
 
+  private uuidsFromHrefSubstringSelector(selector: MemoizedSelector<any, IndexState>, name: string, href: string): MemoizedSelector<any, string[]> {
+    return createSelector(selector, (state: IndexState) => this.getUuidsFromHrefSubstring(state, name, href));
+  }
+
+  private getUuidsFromHrefSubstring(state: IndexState, name: string, href: string): string[] {
```
Comment (tdonohue) [inline src/app/core/data/request.service.ts]: Could we add typedocs or inline comments to this private method. It's not entirely clear to me how this is parsing out the UUIDs.

File: src/app/core/index/index.reducer.ts [hunk 11]
```diff
@@ -60,3 +64,16 @@ function removeFromIndexByValue(state: IndexState, action: RemoveFromIndexByValu
     [action.payload.name]: newSubState
   });
 }
+
+function removeFromIndexBySubstring(state: IndexState, action: RemoveFromIndexByValueAction): IndexState {
```
Comment (tdonohue) [inline src/app/core/index/index.reducer.ts]: Typedocs please

File: src/app/core/metadata/normalized-metadata-schema.model.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
+import { autoserialize } from 'cerialize';
+import { NormalizedObject } from '../cache/models/normalized-object.model';
+import { mapsTo } from '../cache/builders/build-decorators';
+import { CacheableObject } from '../cache/object-cache.reducer';
+import { ListableObject } from '../../shared/object-collection/shared/listable-object.model';
+import { MetadataSchema } from './metadataschema.model';
+
+@mapsTo(MetadataSchema)
+export class NormalizedMetadataSchema extends NormalizedObject implements CacheableObject, ListableObject {
```
Comment (tdonohue) [inline src/app/core/metadata/normalized-metadata-schema.model.ts]: Typedocs please

File: src/app/core/registry/registry.service.ts [hunk 13]
```diff
@@ -220,4 +258,224 @@ export class RegistryService {
...
+
+  public deselectAllMetadataField() {
+    this.store.dispatch(new MetadataRegistryDeselectAllFieldAction())
+  }
+
+  public getSelectedMetadataFields(): Observable<MetadataField[]> {
+    return this.store.pipe(select(selectedMetadataFieldsSelector));
+  }
+
+  public createOrUpdateMetadataSchema(schema: MetadataSchema): Observable<MetadataSchema> {
```
Comment (tdonohue) [inline src/app/core/registry/registry.service.ts]: Large methods like this one need either some detailed Typedocs or (more) inline comments. (It's good to see a few small inline comments already, but I'd recommend adding a few more, or adding TypeDocs)

File: src/app/core/registry/registry.service.ts [hunk 14]
```diff
@@ -220,4 +258,224 @@ export class RegistryService {
...
+    return this.delete(this.metadataSchemasPath, id);
+  }
+
+  public clearMetadataSchemaRequests(): Observable<string> {
+    return this.halService.getEndpoint(this.metadataSchemasPath).pipe(
+      tap((href: string) => this.requestService.removeByHrefSubstring(href))
+    )
+  }
+
+  public createOrUpdateMetadataField(field: MetadataField): Observable<MetadataField> {
```
Comment (tdonohue) [inline src/app/core/registry/registry.service.ts]: Another large method without Typedocs.  (Technically all the new methods in this file should have Typedocs, but as most are one-liners, we could let them slide)

File: src/app/+admin/admin-registries/metadata-registry/metadata-registry.reducers.ts (hunk 3 above)
Comment (artlowel) [reply src/app/+admin/admin-registries/metadata-registry/metadata-registry.reducers.ts]: Added

File: src/app/+admin/admin-registries/metadata-schema/metadata-field-form/metadata-field-form.component.ts (hunk 4 above)
Comment (artlowel) [reply src/app/+admin/admin-registries/metadata-schema/metadata-field-form/metadata-field-form.component.ts]: Added

File: src/app/core/cache/response.models.ts (hunk 5 above)
Comment (artlowel) [reply src/app/core/cache/response.models.ts]: Added

File: src/app/core/data/metadata-schema-data.service.ts (hunk 6 above)
Comment (artlowel) [reply src/app/core/data/metadata-schema-data.service.ts]: Added

File: src/app/core/data/request.models.ts (hunk 8 above)
Comment (artlowel) [reply src/app/core/data/request.models.ts]: Added

File: src/app/core/data/request.reducer.ts (hunk 9 above)
Comment (artlowel) [reply src/app/core/data/request.reducer.ts]: Added

File: src/app/core/data/request.service.ts (hunk 10 above)
Comment (artlowel) [reply src/app/core/data/request.service.ts]: Added

**Comment by LotteHofstede** [comment]
@tdonohue Because my next PR #362 already solves the issue with the warnings you experienced , I simply created an issue #363 to make sure it won't be forgotten about.

File: src/app/core/data/request.service.ts (hunk 10 above)
Comment (tdonohue) [reply src/app/core/data/request.service.ts]: @artlowel : It looks like these typedocs/comments were not yet added.  I see new comments on a different private method here, but not on the `getUuidsFromHrefSubstring()` method.

**Pull request merged**